File: src/Flux.Data/Models/BankAccount.cs
namespace Flux.Data.Models;

public class BankAccount
{
    public int Id { get; set; }

    public Guid OwnerUserId { get; set; }

    public string AccountName { get; set; } = string.Empty;

    public string Owner { get; set; } = string.Empty;

    public decimal Balance { get; set; }

    public AccountType Type { get; set; }

    public decimal? CreditCardAprPercent { get; set; }

    public decimal? SavingsApyPercent { get; set; }

    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;
}

public enum AccountType
{
    Checking,
    Savings,
    CreditCard
}

// What other properties would you add here?

File: src/Flux.Web/src/app/models/bank-account.ts
export interface BankAccount {
  id: number;
  accountName: string;
  owner: string;
  balance: number;
  type: AccountType;
  creditCardAprPercent?: number | null;
  savingsApyPercent?: number | null;
  createdAt: string;
  updatedAt: string;
}

export enum AccountType {
  Checking = 0,
  Savings = 1,
  CreditCard = 2
}

export interface CompoundingProjection {
  name: string;
  periodsPerYear: number;
  annualInterestEarned: number;
  endingBalance: number;
}

export interface CreditCardRateAnalytics {
  accountId: number;
  accountName: string;
  balance: number;
  aprPercent: number;
  effectiveDailyRatePercent: number;
  estimatedMonthlyInterest: number;
  minimumPaymentAmount: number;
  estimatedPayoffMonths: number | null;
  aprRank: number;
}

export interface SavingsRateAnalytics {
  accountId: number;
  accountName: string;
  balance: number;
  apyPercent: number;
  projectedMonthlyInterest: number;
  projectedAnnualInterest: number;
  compoundingScenarios: CompoundingProjection[];
  apyRank: number;
}

export interface CreditCardPortfolioSummary {
  cardCount: number;
  totalBalance: number;
  averageAprPercent: number;
  totalEstimatedMonthlyInterest: number;
}

export interface SavingsPortfolioSummary {
  accountCount: number;
  totalBalance: number;
  averageApyPercent: number;
  totalProjectedMonthlyInterest: number;
  totalProjectedAnnualInterest: number;
}

export interface PortfolioRateAnalyticsResponse {
  creditCards: CreditCardRateAnalytics[];
  creditCardSummary: CreditCardPortfolioSummary;
  savingsAccounts: SavingsRateAnalytics[];
  savingsSummary: SavingsPortfolioSummary;
}

export interface AccountRateAnalyticsResponse {
  accountId: number;
  accountName: string;
  accountType: AccountType;
  creditCard: CreditCardRateAnalytics | null;
  savings: SavingsRateAnalytics | null;
}
